src/pykeen/nn/modules.py

```diff
@@ -148,6 +150,43 @@ class Interaction(nn.Module, Generic[HeadRepresentation, RelationRepresentation,
     #: The symbolic shapes for relation representations
     relation_shape: Sequence[str] = ("d",)
 
+    # if the interaction function's head parameter should only receive a subset of entity representations
+    _head_indices: ClassVar[Optional[Sequence[int]]] = None
+
+    # if the interaction function's tail parameter should only receive a subset of entity representations
+    _tail_indices: ClassVar[Optional[Sequence[int]]] = None
+
+    @classmethod
+    def head_indices(cls) -> Sequence[int]:
+        """Return the entity representation indices used for the head representations."""
+        if cls._head_indices is None:
+            return list(range(len(cls.entity_shape)))
+        return cls._head_indices
+
+    @classmethod
+    def tail_indices(cls) -> Sequence[int]:
+        """Return the entity representation indices used for the tail representations."""
+        if cls._tail_indices is None:
+            tail_entity_shape = cls.entity_shape if cls.tail_entity_shape is None else cls.tail_entity_shape
+            return list(range(len(tail_entity_shape)))
+        return cls._tail_indices
+
+    @classmethod
+    def full_entity_shapes(cls) -> Sequence[str]:
```

Comment:
if this value is constant for the lifetime of the class, maybe it should be calculated in `__init_subclass__`? Though that has a bit of a problem with deep inheritance hierarchies. Maybe it can be cached, since this gets called quite a bit

Reply:
oh nvm this is only called once.